timemachine/cpp/src/kernels/k_exchange.cuh

```diff
@@ -78,4 +108,83 @@ void __global__ k_store_accepted_log_probability(
     }
 }
 
+template <typename RealType, bool Negated>
+void __global__ k_adjust_weights(
+    const int N,
+    const int num_target_mols,
+    const int mol_size,
+    const int *__restrict__ mol_atoms_idxs,
+    const int *__restrict__ mol_offsets,
+    const RealType *__restrict__ per_atom_energies,
+    const RealType inv_kT, // 1 / kT
+    RealType *__restrict__ log_weights) {
+    int mol_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    while (mol_idx < num_target_mols) {
+
+        RealType current_log_weight = log_weights[mol_idx];
+        __int128 weight_accumulator = 0;
+
+        int mol_start = mol_offsets[mol_idx];
+        int mol_end = mol_offsets[mol_idx + 1];
+        int min_atom_idx = mol_atoms_idxs[mol_start];
+        int max_atom_idx = mol_atoms_idxs[mol_end - 1];
+
+        // A loop that in the case of water will be 3x3
+        for (int i = 0; i < mol_size; i++) {
```

Comment:
(not important for this PR, but mol size may also be 1 later on when doing counter-ion sampling).

Reply:
We will have to do more work to support mixed size of mols as you need to know the size of all preceding mols. The same goes for the reference, so left to only handle cases where all of the mols are the same size